### PR TITLE
Feature/implement colorconvert element

### PR DESCRIPTION
**Pull request by rrcarlosrodriguez**

### CODE REVIEW

File: ext/tiovx/gsttiovxcolorconvert.c [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+      break;
+  }
+
+  return target_name;
+}
+
+/* TODO: Move this function to an util */
+static enum vx_df_image_e
+map_gst_video_format_to_vx_format (GstVideoFormat gst_format)
+{
```
Comment (emmadrigal) [inline ext/tiovx/gsttiovxcolorconvert.c:185]: a `gsttiovxutils` already exists on develop, you can move it there

File: ext/tiovx/gsttiovxcolorconvert.c [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+      "tiovx-colorconvert", 0, "TIOVX ColorConvert element");
+}
+
+/* Initialize the new element
+ * Initialize instance structure
+ */
+static void
+gst_tiovx_color_convert_init (GstTIOVXColorconvert * self)
+{
+  self->obj = g_malloc (sizeof (ColorConvertObj));
```
Comment (emmadrigal) [inline ext/tiovx/gsttiovxcolorconvert.c:286]: Use `g_malloc0` this ensures 0 initialized memory

File: ext/tiovx/gsttiovxcolorconvert.c [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+    GST_PAD_SRC,
+    GST_PAD_ALWAYS,
+    GST_STATIC_CAPS (TIOVX_COLOR_CONVERT_STATIC_CAPS_SRC)
+    );
+
+struct _GstTIOVXColorconvert
+{
+  GstTIOVXSiso element;
+  gint target_id;
+  ColorConvertObj *obj;
```
Comment (emmadrigal) [inline ext/tiovx/gsttiovxcolorconvert.c:129]: Why a pointer?

From it's intialization we are just using malloc, no specitial requirements should prevent us from using:
`ColorconvertObj obj;`

File: ext/tiovx/gsttiovxcolorconvert.c [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+}
+
+static void
+gst_tiovx_color_convert_dispose (GObject * obj)
+{
+  GstTIOVXColorconvert *self = GST_TIOVX_COLOR_CONVERT (obj);
+
+  if (self->obj) {
+    g_free (self->obj);
+  }
```
Comment (emmadrigal) [inline ext/tiovx/gsttiovxcolorconvert.c:445]: g_free simply returns if its argument is `NULL` there is no need to check for non-NULL before calling it

File: ext/tiovx/gsttiovxcolorconvert.c [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+/* GstTIOVXSiso Functions */
+
+static gboolean
+gst_tiovx_color_convert_init_module (GstTIOVXSiso * trans, vx_context context,
+    GstVideoInfo * in_info, GstVideoInfo * out_info, guint in_pool_size,
+    guint out_pool_size)
+{
+  GstTIOVXColorconvert *self = GST_TIOVX_COLOR_CONVERT (trans);
+  vx_status status = VX_SUCCESS;
+  ColorConvertObj *colorconvert;
```
Comment (emmadrigal) [inline ext/tiovx/gsttiovxcolorconvert.c:459]: Initialize to NULL

**Review by emmadrigal — CHANGES_REQUESTED** [review]

**Review by emmadrigal — APPROVED** [review]

File: gst-libs/gst/tiovx/gsttiovxsiso.c [hunk 6]
```diff
@@ -302,14 +282,17 @@ gst_tiovx_siso_finalize (GObject * obj)
...
+  }
 
   /* App common deinit */
   GST_DEBUG_OBJECT (self, "Running TIOVX common deinit");
-  tivxHostDeInit ();
-  tivxDeInit ();
-  appCommonDeInit ();
+  if (priv->tiovx_context) {
+    g_object_unref (priv->tiovx_context);
+  }
```
Comment (jafet-chaves) [inline gst-libs/gst/tiovx/gsttiovxsiso.c:295]: Should not this function chain the gstelementclass finalize method?

File: gst-libs/gst/tiovx/gsttiovxsiso.c [hunk 7]
```diff
@@ -190,33 +187,16 @@ gst_tiovx_siso_init (GstTIOVXSiso * self)
...
   /* App common init */
   GST_DEBUG_OBJECT (self, "Running TIOVX common init");
-  if (0 != appCommonInit ()) {
-    GST_ERROR_OBJECT (self, "App common init failed");
-    goto exit;
-  }
-
-  tivxInit ();
-  tivxHostInit ();
+  priv->tiovx_context = gst_tiovx_context_new ();
```
Comment (jafet-chaves) [inline gst-libs/gst/tiovx/gsttiovxsiso.c:190]: We should check if tiovx_context is not NULL, also should be initialized before

File: ext/tiovx/gsttiovxcolorconvert.c [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+
+/* Target definition */
+#define GST_TYPE_TIOVX_COLOR_CONVERT_TARGET (gst_tiovx_color_convert_target_get_type())
+static GType
+gst_tiovx_color_convert_target_get_type (void)
+{
+  static GType target_type = 0;
+
+  static const GEnumValue targets[] = {
+    {TIVX_CPU_ID_DSP1, "DSP1", "dsp1"},
```
Comment (michaelgruner) [inline ext/tiovx/gsttiovxcolorconvert.c:82]: The enum is formed by three members:
1. The numeric value
2. A human readable description
3. A short string tag.

Use [TI string targets](http://software-dl.ti.com/jacinto7/esd/processor-sdk-rtos-jacinto7/latest/exports/docs/tiovx/docs/user_guide/group__group__tivx__ext__targets.html) for 3 and later you can use:

```
GType type;
GEnumClass *enum_class;
GEnumValue *enum_value;
  
type = gst_tiovx_color_convert_target_get_type ();
enum_class = G_ENUM_CLASS (type);
enum_value = g_enum_get_value (enum_class, self->target_id);
target_name = enum_value->nick
```

GEnumValue * enum = g_enum_get_value (GEnumClass *enum_class,
                  gint value);

File: gst-libs/gst/tiovx/gsttiovxsiso.c [hunk 9]
```diff
@@ -596,6 +579,13 @@ gst_tiovx_siso_modules_init (GstTIOVXSiso * self)
     goto exit;
   }
 
+  status = vxGetStatus ((vx_reference) priv->context);
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxsiso.c:582]: Again, feels kind of random the stuff we check. Sometimes we treat them as programming errors, sometimes (as this one) we treat them as runtime errors.

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 10]
```diff
@@ -100,3 +101,64 @@ vx_format_to_gst_format (const vx_df_image format)
...
+    case TIVX_CPU_ID_DSP2:
+      target_name = TIVX_TARGET_DSP2;
+      break;
+    default:
+      target_name = NULL;
+      break;
+  }
+
+  return target_name;
+}
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxutils.c:164]: Not needed anymore, we should take the name out of the enum.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

**Pull request merged**